File: packages/tools/src/tools/write.test.ts
import { test } from "node:test";
import { strictEqual, ok } from "node:assert";
import { readFile, mkdir, rm, access } from "node:fs/promises";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { writeTool } from "./write.js";
import type { ToolContext } from "@forge/types";

const testDir = join(tmpdir(), `forge-write-test-${Date.now()}`);

// Helper to create test context
function makeCtx(): ToolContext {
  return {
    cwd: testDir,
    sessionId: "test-session",
    threadId: "test-thread",
    signal: new AbortController().signal,
    emit: () => {},
  };
}

test("Write: creates new file with content", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "new-file.txt");

  const result = await writeTool.handler(
    { file_path: file, content: "Hello, Troy Barnes!" },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);
  strictEqual(result.content.length, 1);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("19 bytes"));

  // Verify file was actually created
  const contents = await readFile(file, "utf-8");
  strictEqual(contents, "Hello, Troy Barnes!");

  await rm(testDir, { recursive: true });
});

test("Write: creates parent directories", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "deep", "nested", "dir", "file.txt");

  const result = await writeTool.handler(
    { file_path: file, content: "Greendale Community College" },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("27 bytes"));

  // Verify file exists
  const contents = await readFile(file, "utf-8");
  strictEqual(contents, "Greendale Community College");

  await rm(testDir, { recursive: true });
});

test("Write: overwrites existing file", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "overwrite.txt");

  // Write initial content
  await writeTool.handler(
    { file_path: file, content: "Original content" },
    makeCtx(),
  );

  // Overwrite
  const result = await writeTool.handler(
    { file_path: file, content: "New content from Señor Chang" },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("29 bytes"));

  // Verify new content
  const contents = await readFile(file, "utf-8");
  strictEqual(contents, "New content from Señor Chang");

  await rm(testDir, { recursive: true });
});

test("Write: handles empty content", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "empty.txt");

  const result = await writeTool.handler(
    { file_path: file, content: "" },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("0 bytes"));

  // Verify file exists and is empty
  const contents = await readFile(file, "utf-8");
  strictEqual(contents, "");

  await rm(testDir, { recursive: true });
});

test("Write: returns error on write failure", async () => {
  // Try to write to a path that can't be created (e.g., root without permissions)
  const invalidPath = "/this/path/should/not/be/writable/test.txt";

  const result = await writeTool.handler(
    { file_path: invalidPath, content: "test" },
    makeCtx(),
  );

  strictEqual(result.isError, true);
  ok(result.content[0].type === "text");
  ok(
    result.content[0].text.includes("Failed to write") ||
      result.content[0].text.includes("EACCES") ||
      result.content[0].text.includes("EPERM"),
  );
});

File: internal/tools/write_test.go
package tools

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/jelmersnoeck/forge/internal/types"
	"github.com/stretchr/testify/require"
)

func TestWriteTool(t *testing.T) {
	tests := map[string]struct {
		setup func(t *testing.T, dir string) map[string]any
		want  func(*testing.T, string, types.ToolResult, error)
	}{
		"write new file": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "troy.txt")
				return map[string]any{
					"file_path": path,
					"content":   "Troy and Abed in the morning",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				r.Contains(result.Content[0].Text, "wrote")

				// Verify file was created
				path := filepath.Join(dir, "troy.txt")
				data, readErr := os.ReadFile(path)
				r.NoError(readErr)
				r.Equal("Troy and Abed in the morning", string(data))
			},
		},
		"write with nested directories": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "greendale", "study_room", "notes.txt")
				return map[string]any{
					"file_path": path,
					"content":   "Señor Chang was here",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)

				// Verify nested file was created
				path := filepath.Join(dir, "greendale", "study_room", "notes.txt")
				data, readErr := os.ReadFile(path)
				r.NoError(readErr)
				r.Equal("Señor Chang was here", string(data))
			},
		},
		"overwrite existing file": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "dean.txt")
				err := os.WriteFile(path, []byte("Old content"), 0644)
				require.NoError(t, err)
				return map[string]any{
					"file_path": path,
					"content":   "Dean Pelton approves",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)

				// Verify file was overwritten
				path := filepath.Join(dir, "dean.txt")
				data, readErr := os.ReadFile(path)
				r.NoError(readErr)
				r.Equal("Dean Pelton approves", string(data))
			},
		},
		"empty content": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "empty.txt")
				return map[string]any{
					"file_path": path,
					"content":   "",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)

				// Verify empty file was created
				path := filepath.Join(dir, "empty.txt")
				data, readErr := os.ReadFile(path)
				r.NoError(readErr)
				r.Len(data, 0)
			},
		},
		"missing file_path": {
			setup: func(t *testing.T, dir string) map[string]any {
				return map[string]any{"content": "some content"}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.Error(err)
				r.True(result.IsError)
			},
		},
		"missing content": {
			setup: func(t *testing.T, dir string) map[string]any {
				return map[string]any{"file_path": filepath.Join(dir, "test.txt")}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.Error(err)
				r.True(result.IsError)
			},
		},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			dir := t.TempDir()
			input := tc.setup(t, dir)
			tool := WriteTool()
			result, err := tool.Handler(input, types.ToolContext{})
			tc.want(t, dir, result, err)
		})
	}
}
